### PR TITLE
Add feature check

### PR DESCRIPTION
**Pull request by HelloHorizon**

### CODE REVIEW

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        .equals(ServerConfiguration.get(PropertyKey.MASTER_JOURNAL_TYPE));
+  }
+
+  /**
+   * Utility to check the master metastore is rocks.
+   *
+   * @return true, if running with rocks
+   */
+  public static boolean isRocks() {
+    return "ROCKS".equals(ServerConfiguration.get(PropertyKey.MASTER_METASTORE));
```
Comment (Xenorith) [inline core/server/common/src/main/java/alluxio/util/FeatureUtils.java:38]: can this be `MetastoreType.ROCKS.toString()`?

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 2]
```diff
@@ -98,7 +111,46 @@ public static String getUserAgentString(String clusterID) throws IOException {
...
+      featureInfo.add("persistBlackList");
+    }
+    if (FeatureUtils.isUnsafeDirectPersistEnable()) {
+      featureInfo.add("unsafePersist");
+    }
+    if (FeatureUtils.isMasterAuditLoggingEnable()) {
+      featureInfo.add("masterAuditLog");
+    }
+    if (featureInfo.size() == 0) {
+      return null;
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:151]: should the method be annotated somehow if this method can return null?
is there a unit test for this corner case?

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 3]
```diff
@@ -84,6 +85,18 @@ public static String getLatestVersion(String clusterID, long connectionRequestTi
    */
   @VisibleForTesting
   public static String getUserAgentString(String clusterID) throws IOException {
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:87]: update javadoc for this method, specifically the `@return`

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 4]
```diff
@@ -84,6 +85,18 @@ public static String getLatestVersion(String clusterID, long connectionRequestTi
...
+    Joiner joiner = Joiner.on("; ").skipNulls();
+    String sysInfo = joiner.join(getUserAgentEnvironmentString(clusterID), getFeatureString());
+    return String.format("Alluxio/%s (%s)", ProjectConstants.VERSION, sysInfo);
+  }
+
+  /**
+   * @param clusterID the cluster ID
+   * @return a string representation of the user's environment in the format "key1:value1, key2:
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:95]: is this javadoc accurate? i don't think they are always key/value pairs and each section is separated by `;`, not `,`

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 5]
```diff
@@ -98,7 +111,46 @@ public static String getUserAgentString(String clusterID) throws IOException {
...
       }
     }
-    return String.format("Alluxio/%s (%s)", ProjectConstants.VERSION, sysInfo);
+    return sysInfo;
+  }
+
+  /**
+   * Get the features information.
+   *
+   * @return a list of string represent feature names
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:120]: `a list of strings representing enabled features in the format ...`

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 6]
```diff
@@ -46,86 +53,135 @@ public void before() {
...
     Mockito.when(EnvironmentUtils.getEC2ProductCode()).thenReturn("random123code");
     Mockito.when(EnvironmentUtils.isEMR(Mockito.anyString())).thenReturn(true);
     Mockito.when(EC2MetadataUtils.getUserData()).thenReturn("emr_apps");
 
+    String userAgentString = UpdateCheck.getUserAgentEnvironmentString("cluster1");
+    Assert.assertTrue(userAgentString.equals("cluster1; ProductCode:random123code; emr; ec2"));
+  }
+
+  @Test
+  public void featureString() {
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:129]: should this be split up into different unit tests for each individual feature to be consistent with the existing unit tests?

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 7]
```diff
@@ -46,86 +53,135 @@ public void before() {
...
+    // Embedded journal
+    ServerConfiguration.set(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.UFS);
+    Assert.assertTrue(!UpdateCheck.getFeatureString().contains("embedded"));
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:132]: using `String.contains` is a weak check since another feature could have the substring `embedded` or have the equivalent value. how can this unit test be reformatted to catch the bad scenario where there's another feature that is also called `embedded`? same for other tests

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 8]
```diff
@@ -46,86 +53,135 @@ public void before() {
...
+    ServerConfiguration.set(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.EMBEDDED);
+    Assert.assertTrue(UpdateCheck.getFeatureString().contains("embedded"));
+
+    // Rocks
+    ServerConfiguration.set(PropertyKey.MASTER_METASTORE, MetastoreType.ROCKS);
+    Assert.assertTrue(UpdateCheck.getFeatureString().contains("rocks"));
+    ServerConfiguration.set(PropertyKey.MASTER_METASTORE, MetastoreType.HEAP);
+    Assert.assertTrue(!UpdateCheck.getFeatureString().contains("rocks"));
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:140]: `Assert.assertTrue(!...)` should be `Assert.assertFalse(...)`

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  /**
+   * Utility to check Zookeeper is enabled.
+   *
+   * @return true, if Zookeeper is enabled
+   */
+  public static boolean isZookeeperEnable() {
```
Comment (Xenorith) [inline core/server/common/src/main/java/alluxio/util/FeatureUtils.java:46]: instead of having a method for each property, can we have a single method for boolean properties that takes in the argument for the property key? this would make it easier to reuse in the future

Comment (Xenorith) [inline core/server/common/src/main/java/alluxio/util/FeatureUtils.java:46]: isZookeeperEnable -> isZookeeperEnabled, similar for below

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    return ServerConfiguration.getBoolean(PropertyKey.MASTER_DAILY_BACKUP_ENABLED);
+  }
+
+  /**
+   * Utility to check persistence black list is empty.
+   *
+   * @return true, if persistence black list is empty
+   */
+  public static boolean isPersistenceBlacklistEmpty() {
+    if (ServerConfiguration.isSet(PropertyKey.MASTER_PERSISTENCE_BLACKLIST)) {
```
Comment (Xenorith) [inline core/server/common/src/main/java/alluxio/util/FeatureUtils.java:74]: why split this up into 2 nested if statements? seems like it can be represented as a single line of `return condition1 &&/|| condition2`

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 11]
```diff
@@ -98,7 +111,46 @@ public static String getUserAgentString(String clusterID) throws IOException {
...
+   */
+  @VisibleForTesting
+  public static String getFeatureString() {
+    Joiner joiner = Joiner.on("; ").skipNulls();
+    List<String> featureInfo = new ArrayList<>();
+    if (FeatureUtils.isEmbeddedJournal()) {
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:126]: this list looks really long and repetitive. not sure if this is possible in java, but could this be represented as a map of string to function, with a loop iterating over the map?

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 12]
```diff
@@ -46,86 +53,135 @@ public void before() {
...
+    Assert.assertTrue(UpdateCheck.getFeatureString().contains("masterAuditLog"));
+    ServerConfiguration.set(PropertyKey.MASTER_AUDIT_LOGGING_ENABLED, false);
+    Assert.assertTrue(!UpdateCheck.getFeatureString().contains("masterAuditLog"));
+  }
+
+  @Test
+  public void userAgent() throws Exception {
     String userAgentString = UpdateCheck.getUserAgentString("cluster1");
-    Assert.assertTrue(userAgentString.equals(
-        String.format("Alluxio/%s (cluster1; ProductCode:random123code; emr; ec2)",
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:128]: is this not how the string ends up being anymore? i prefer the strict equals check than a regex check to be able to potentially catch misformatted entries

**Review by Xenorith — CHANGES_REQUESTED** [review]

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java (hunk 10 above)
Comment (HelloHorizon) [reply core/server/common/src/main/java/alluxio/util/FeatureUtils.java]: addressed

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java (hunk 4 above)
Comment (HelloHorizon) [reply core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: addressed

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java (hunk 8 above)
Comment (HelloHorizon) [reply core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java]: addressed

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java (hunk 6 above)
Comment (HelloHorizon) [reply core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java]: addressed

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 13]
```diff
@@ -46,86 +53,148 @@ public void before() {
...
+  public void featureStringEmbeddedJournal() {
+    ServerConfiguration.set(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.UFS);
+    Assert.assertFalse(UpdateCheck.getFeatureString().contains("embedded"));
+    ServerConfiguration.set(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.EMBEDDED);
+    Assert.assertTrue(UpdateCheck.getFeatureString().contains("embedded"));
+  }
+
+  @Test
+  public void featureStringRocks() {
+    ServerConfiguration.set(PropertyKey.MASTER_METASTORE, MetastoreType.ROCKS);
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:138]: i'm confused why you can use MetastoreType.ROCKS here in the test file but not the source file, is this expected?

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java (hunk 9 above)
Comment (Xenorith) [reply core/server/common/src/main/java/alluxio/util/FeatureUtils.java]: sorry, i dont understand the reasoning behind why we shouldn't be more efficient in writing methods in utils class?
what is the issue with having:
```
public static boolean isBooleanPropertyEnabled(String propertyKey) {
  return ServerConfiguration.getBoolean(propertyKey);
}
```
to replace all of the simple boolean check methods?

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java [hunk 14]
```diff
@@ -46,86 +53,148 @@ public void before() {
...
     Mockito.when(EC2MetadataUtils.getUserData()).thenReturn("emr_apps");
 
+    String userAgentString = UpdateCheck.getUserAgentEnvironmentString("cluster1");
+    Assert.assertTrue(userAgentString.equals("cluster1; ProductCode:random123code; emr; ec2"));
+  }
+
+  @Test
+  public void featureStringEmbeddedJournal() {
+    ServerConfiguration.set(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.UFS);
+    Assert.assertFalse(UpdateCheck.getFeatureString().contains("embedded"));
```
Comment (Xenorith) [inline core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:131]: readding previous comment that got marked as outdated but was not addressed. this is especially important if the `userAgent()` unit test at the bottom of this file is going to do a regex check instead of an exact match.

using `String.contains` is a weak check since another feature could have the substring embedded or have the equivalent value. how can this unit test be reformatted to catch the bad scenario where there's another feature that is also called `embedded`? same for other tests

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 15]
```diff
@@ -79,11 +80,22 @@ public static String getLatestVersion(String clusterID, long connectionRequestTi
...
    * @param clusterID the cluster ID
-   * @return a string representation of the user's environment in the format "key1:value1, key2:
-   *         value2".
+   * @return a string representation of the user's environment in the format
+   *         "Alluxio/{ALLUXIO_VERSION} (valueA; valueB)"
    */
   @VisibleForTesting
   public static String getUserAgentString(String clusterID) throws IOException {
+    Joiner joiner = Joiner.on("; ").skipNulls();
+    String sysInfo = joiner.join(getUserAgentEnvironmentString(clusterID), getFeatureString());
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:89]: `getUserAgentEnvironmentString()` always returns a non-empty string but i believe we want `getFeatureString` to return null if there are zero features enabled. correct me if i'm wrong, but a joiner on no elements is going to return `""` and so you will end up joining `"some env string"` with `""` -> `"some env string; "` instead of `"some env string"`

regardless of the behavior of joiner, we should have a unit test for this case as i had asked for previously

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 16]
```diff
@@ -79,11 +80,22 @@ public static String getLatestVersion(String clusterID, long connectionRequestTi
...
+    return String.format("Alluxio/%s (%s)", ProjectConstants.VERSION, sysInfo);
+  }
+
+  /**
+   * @param clusterID the cluster ID
+   * @return a string representation of the user's environment in the format "docker; kubernetes"
+   */
+  @VisibleForTesting
+  public static String getUserAgentEnvironmentString(String clusterID) throws IOException {
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:98]: either this method should be named `getEnvironmentString` or the below method should be `getUserAgentFeatureString` to be consistent with each other

**Review by Xenorith — CHANGES_REQUESTED** [review]

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java (hunk 13 above)
Comment (HelloHorizon) [reply core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:138]: It's because of the dependency of different modules. Both `UpdateCheckTest` and  `JournalType" are under alluxio-core-server-common, while `MetastoreType` is under alluxio-core-server-master. If we want to reuse the enum MetastoreType, we have to make alluxio-core-server-master as a dependency as alluxio-core-server-common, which is not expected. The workaround is using string instead, and check consistency in a unit test.

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java (hunk 9 above)
Comment (HelloHorizon) [reply core/server/common/src/main/java/alluxio/util/FeatureUtils.java]: I am not sure I understand the difference between using the existing method compare to change to `isBooleanPropertyEnabled`. If we want to abstract a method like `isBooleanPropertyEnabled`, we can directly use `ServerConfiguration.getBoolean(propertyKey)` instead.

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java (hunk 13 above)
Comment (Xenorith) [reply core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:138]: does the test module have different dependencies than the source module? because you are importing `import alluxio.master.metastore.MetastoreType;` in this test class. again, my confusion is why you can import it here in the test class but not in the utils class

File: core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java (hunk 14 above)
Comment (HelloHorizon) [reply core/server/master/src/test/java/alluxio/master/meta/UpdateCheckTest.java:131]: Since if we have more fields, we have to mute all configurations one by one before the test. Just let me know if you have a better idea to do these tests

File: core/server/common/src/main/java/alluxio/util/FeatureUtils.java (hunk 9 above)
Comment (Xenorith) [reply core/server/common/src/main/java/alluxio/util/FeatureUtils.java]: honestly that sounds like a better idea than mine

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 17]
```diff
@@ -98,7 +114,39 @@ public static String getUserAgentString(String clusterID) throws IOException {
...
+    return sysInfo;
+  }
+
+  /**
+   * Get the features information.
+   *
+   * @return a list of strings representing enabled features
+   */
+  @VisibleForTesting
+  public static List<String> getUserAgentFeatureString() {
```
Comment (Xenorith) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java:126]: since this doesn't return a String anymore, rename to `getUserAgentFeatures()`?

**Review by Xenorith — APPROVED** [review]
looks good to me other than considering a small rename

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java (hunk 17 above)
Comment (HelloHorizon) [reply core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: addressed

**Comment by HelloHorizon** [comment]
alluxio-bot, merge this please
